Add input_boolean reload service

Adminiuga opened the pull request:
## Description:
Implements `input_boolean.reload` service. There's no reason to restart the entire HA instance, only to add or remove a few boolean inputs which I use for automations.

If this is a correct approach, I'll implement similar changes for the rest of the input platforms unless the goal is to move these into config entries. In this case I'd need an example platform as a model. Zones???

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.
  - [x] I have followed the [development checklist][dev-checklist]

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] [_The manifest file_][manifest-docs] has all fields filled out correctly. Update and include derived files by running `python3 -m script.hassfest`.
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `python3 -m script.gen_requirements_all`.
  - [x] Untested files have been added to `.coveragerc`.

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works.

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html

Adminiuga committed:
Add input_boolean reload service.

balloob reviewed (approved):
This is great. Ok to merge when test added 👍

Adminiuga committed:
Add reload test.

Adminiuga commented:
Any hints how to modify `config[DOMAIN]` in  test after component was loaded via `await async_setup_component()` ?

Adminiuga commented:
```
new_config = {
  "input_boolean": {
    "ib1": {"name": "input 1","initial": True},
    "ib2": None,
  }
}

await homeassistant.config.merge_packages_config(hass, config, {})
```

can I just do?

balloob commented:
Check how we do it for groups: https://github.com/home-assistant/home-assistant/blob/52926f19618ce8049bb184c30a839a5c04f479a1/tests/components/group/test_init.py#L421-L435

Adminiuga commented:
Added test. I guess it is ready for  the final review?

balloob commented on homeassistant/components/input_boolean/__init__.py:
```diff
@@ -52,6 +55,35 @@ def is_on(hass, entity_id):
 
+    entities = await _async_process_config(config)
+    if not entities:
+        return False
```
Let's not return false. It is totally fine to set up the integration without entities. We can reload and add more later !

balloob replied:
I would just not call `component.async_add_entities`, still register service.

Adminiuga replied:
addressed in a31b0eb391bbb7222c9eb9ffaa931d30484908b3

balloob commented on homeassistant/components/input_boolean/__init__.py:
```diff
@@ -52,6 +55,35 @@ def is_on(hass, entity_id):
+        if new_entities:
+            await component.async_add_entities(new_entities)
+
+    hass.services.async_register(
```
Please use `helpers.service.async_register_admin_service`

Adminiuga replied:
Makes sense. I copy'n'pasted from `automation` platform. Do we want to register `automation.reload` service as an admin service too?

Adminiuga replied:
Addressed in a31b0eb391bbb7222c9eb9ffaa931d30484908b3

balloob reviewed (approved):
Ok to merge when final 2 comments addressed

Adminiuga committed:
Address comments.

Register reload service as an admin service
Setup platform even if there're no entities

Adminiuga commented:
Addressed comments and ready for the Final-Final review...
